chain/db/pubsub.go

```diff
@@ -2,95 +2,220 @@ package db
-
-	return topicKey.Bytes()
+type pubSub struct {
+	subscriber string
+	meta       string
+	expiresAt  uint32
 }
 
-func generateTopicBucketKey(topic string, bucket uint32) []byte {
-	topicBucketKey := bytes.NewBuffer(nil)
-	topicKey := generateTopicKey(topic)
-	topicBucketKey.Write(topicKey)
-	serialization.WriteUint32(topicBucketKey, bucket)
-
-	return topicBucketKey.Bytes()
-}
+func (ps *pubSub) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarString(w, ps.subscriber); err != nil {
+		return err
+	}
+	if err := serialization.WriteVarString(w, ps.meta); err != nil {
+		return err
+	}
 
-func generateSubscriberKey(subscriber []byte, identifier string, topic string, bucket uint32) []byte {
-	subscriberKey := bytes.NewBuffer(nil)
-	topicBucketKey := generateTopicBucketKey(topic, bucket)
-	subscriberKey.Write(topicBucketKey)
-	serialization.WriteVarBytes(subscriberKey, subscriber)
-	serialization.WriteVarString(subscriberKey, identifier)
+	if err := serialization.WriteUint32(w, ps.expiresAt); err != nil {
+		return err
+	}
 
-	return subscriberKey.Bytes()
+	return nil
 }
 
-func (cs *ChainStore) Subscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, meta string, height uint32) error {
-	if duration == 0 {
-		return nil
-	}
+func (ps *pubSub) Deserialize(r io.Reader) error {
+	var err error
 
-	subscriberKey := generateSubscriberKey(subscriber, identifier, topic, bucket)
+	ps.subscriber, err = serialization.ReadVarString(r)
+	if err != nil {
+		return err
+	}
 
-	// PUT VALUE
-	err := cs.st.BatchPut(subscriberKey, []byte(meta))
+	ps.meta, err = serialization.ReadVarString(r)
 	if err != nil {
 		return err
 	}
 
-	err = cs.ExpireKeyAtBlock(height+duration, subscriberKey)
+	ps.expiresAt, err = serialization.ReadUint32(r)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (cs *ChainStore) Unsubscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, height uint32) error {
-	if duration == 0 {
-		return nil
+func (ps *pubSub) Empty() bool {
+	return len(ps.subscriber) == 0 && len(ps.meta) == 0 && ps.expiresAt == 0
+}
+
+func getTopicId(topic string) []byte {
+	topicHash := sha256.Sum256([]byte(strings.ToLower(topic)))
+	return topicHash[:20]
+}
+
+func getTopicBucketId(topic string, bucket uint32) []byte {
+	topicBucketId := bytes.NewBuffer(nil)
+
+	topicId := getTopicId(topic)
+	topicBucketId.Write(topicId)
+
+	serialization.WriteUint32(topicBucketId, bucket)
+
+	return topicBucketId.Bytes()
+}
+
+func getPubSubId(topic string, bucket uint32, subscriber []byte, identifier string) string {
+	pubSubId := bytes.NewBuffer(nil)
+
+	pubSubId.Write(getTopicBucketId(topic, bucket))
+
+	subscriberHash := sha256.Sum256(append(subscriber, identifier...))
+	pubSubId.Write(subscriberHash[:20])
+
+	return string(pubSubId.Bytes())
+}
+
+func getPubSubCleanupId(height uint32) string {
+	buf := new(bytes.Buffer)
+	_ = serialization.WriteUint32(buf, height)
```

Comment:
I think it's better to handle errors in getTopicBucketId, getPubSubId, getPubSubCleanupId (serialization.writeXXX and write) because in theory the error could happen, very unlikely but still possible

Reply:
I don't think error is possible if you look at buffer.go/line 177. it's nil everytime. error could happen depending on Writer, but we user Buffer here, so...